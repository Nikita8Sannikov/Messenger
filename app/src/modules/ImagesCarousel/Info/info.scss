.info {
    position: absolute;
    top: 20px;
    left: 0px;
    width: auto;
    height: auto;
    display: flex;
    align-items: center;
    justify-content: center;
    gap: 5px;

    &__container {
        min-width: max-content;
        display: flex;
        flex-direction: column;
        justify-content: center;
        align-items: flex-start;
    }
}