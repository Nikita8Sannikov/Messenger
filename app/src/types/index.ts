// В данном файле хранятся общие абстрактные типы, применимые на всем уровне клиента

export type ValueOf<T> = T[keyof T];

export type TimeoutType = ReturnType<typeof setTimeout>;
