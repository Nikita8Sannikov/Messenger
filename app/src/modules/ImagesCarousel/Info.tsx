import React from "react";

import { transformDate } from "../../utils/time";
import { useAppSelector } from "../../hooks/useGlobalState";
import { selectUserState } from "../../state/user/slice";
import { getFullName } from "../../utils";
import Avatar from "../../components/Common/Avatar";

import "./info.scss";

export interface ICarouselImage {
    src: string;
    authorName: string;
    dateTime: string;
    authorAvatarUrl: string;
    alt: string;
};

export default React.memo(function Info({ activeImage }: { activeImage: ICarouselImage; }) {
    const [name, setName] = React.useState("");
    const [createDate, setCreateDate] = React.useState("");

    const { user } = useAppSelector(selectUserState);

    React.useEffect(() => {
        setName(user && getFullName(user) === activeImage.authorName
            ? "Вы"
            : activeImage.authorName
        );
    }, [activeImage.authorName]);

    React.useEffect(() => {
        if (activeImage.dateTime) {
            setCreateDate(transformDate(activeImage.dateTime, true));
        }
    }, [activeImage.dateTime]);

    return <div className="info">
        <Avatar src={activeImage.authorAvatarUrl} alt={activeImage.alt} />

        <div className="info__container">
            <div className="info__container__user-name">
                {name}
            </div>

            <div className="info__container__date-time">
                {createDate}
            </div>
        </div>
    </div>
});