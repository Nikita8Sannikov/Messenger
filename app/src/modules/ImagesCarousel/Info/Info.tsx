import { useState, useEffect, memo } from "react";
import { useTranslation } from "react-i18next";

import useUser from "../../../hooks/useUser";
import AvatarComponent from "../../../components/ui/Avatar";
import { transformDate } from "../../../utils/time";

import "./info.scss";

export interface ICarouselImage {
    src: string;
    authorName: string;
    dateTime: string;
    authorAvatarUrl: string;
    alt: string;
};

export default memo(function Info({ activeImage }: { activeImage: ICarouselImage; }) {
    const [name, setName] = useState("");
    const [createDate, setCreateDate] = useState("");

    const { t } = useTranslation();
    const user = useUser();

    useEffect(() => {
        setName(user.fullName === activeImage.authorName
            ? t("images-carousel-module.you")
            : activeImage.authorName
        );
    }, [activeImage.authorName]);

    useEffect(() => {
        if (activeImage.dateTime) {
            setCreateDate(transformDate(activeImage.dateTime, true));
        }
    }, [activeImage.dateTime]);

    return <div className="info">
        <AvatarComponent src={activeImage.authorAvatarUrl} alt={activeImage.alt} />

        <div className="info__container">
            <div className="info__container__user-name">
                {name}
            </div>

            <div className="info__container__date-time">
                {createDate}
            </div>
        </div>
    </div>
});