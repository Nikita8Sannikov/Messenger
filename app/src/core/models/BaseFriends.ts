import { ApiRoutes } from "common-types";

import { type IFriend } from "@custom-types/friends.types";

export interface IGetAll<T> { 
    route: ApiRoutes;
    page?: number;
    data?: { ids: string[] };
    setLoading: (isLoading: boolean) => void; 
    successCb: (data: { success: boolean; } & T) => void; 
};

// Контракт модели "Управление друзьями"
export interface BaseFriends {
    setCount(count: number): void;
    add(user: IFriend): void;
    remove(userId: string): void;
    find(userId: string): IFriend | undefined;
};