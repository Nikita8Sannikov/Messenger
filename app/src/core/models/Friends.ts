import { type FriendsTab } from "common-types";

import type FriendsController from "@core/controllers/FriendsController";
import { type IUser } from "@custom-types/models.types";

// Контракт модели "Сервис друзей"
export interface Friends {
    friendsController: FriendsController;
    
    checkOnlineUsers(users: IUser[]): void;
    removeOnlineUser(userId: string): void;
    getFriends(type: FriendsTab): void;
    loadMore(type: FriendsTab, resolve: () => void): void;
    search(type: FriendsTab, value: string): void;

    followFriend(friendId: string): void;
    writeMessage(friendId: string): void;
    addFriend(friendId: string): void;
    accept(friendId: string): void;
    leftInFollowers(friendId: string): void;
    unfollow(friendId: string): void;
    deleteFriend(friendId: string): void;
    blockFriend(friendId: string): void;
    unblock(friendId: string): void;
};