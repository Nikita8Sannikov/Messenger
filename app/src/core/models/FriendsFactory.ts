import { type BaseFriends } from "@core/models/BaseFriends";

// Контракт модели "Фабрика друзей"
export interface FriendsFactory {
    createFriendsManager(): BaseFriends;
    createOutgoingRequestsManager(): BaseFriends;
    createFollowersManager(): BaseFriends;
    createBlockedUsersManager(): BaseFriends;
    createOnlineFriendsManager(): BaseFriends;
    createCommonFriendsManager(): BaseFriends;
};