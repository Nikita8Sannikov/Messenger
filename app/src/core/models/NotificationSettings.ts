import { type INotificationSettings } from "@custom-types/models.types";

// Контракт модели "Настройки уведомлений пользователя"
export interface NotificationSettings {
	settings: INotificationSettings;
};