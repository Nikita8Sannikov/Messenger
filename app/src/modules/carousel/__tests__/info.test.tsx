import { render, screen } from "@testing-library/react";

import Info from "@modules/carousel/info";
import { type IUser } from "@custom-types/models.types";
import mockUserStore from "../../../__mocks__/@store/user";

jest.mock("@service/i18n", () => ({
	t: (key: string) => {
		const translations: Record<string, string> = {
			"images-carousel-module.you": "You",
			"utils.months.march": "march",
		};
		return translations[key] || key;
	},
}));

describe("Info", () => {
	const mockActiveImage = {
		src: "test-image.jpg",
		alt: "Test Image",
		authorName: "John Doe",
		dateTime: "2024-03-20T10:00:00Z",
		authorAvatarUrl: "avatar.jpg",
	};

	beforeEach(() => {
		mockUserStore.getState().user = {
			id: "123",
			fullName: "John Doe",
		} as IUser;
	});

	it("renders user avatar with correct props", () => {
		render(<Info activeImage={mockActiveImage} />);
		const avatar = screen.getByRole("img", { name: "Test Image" });
    
		expect(avatar).toBeInTheDocument();
		expect(avatar).toHaveAttribute("src", "avatar.jpg");
		expect(avatar).toHaveAttribute("alt", "Test Image");
	});

	it("displays author name when user is not the author", () => {
		mockUserStore.getState().user = {
			id: "456",
			fullName: "Different User",
		} as IUser;

		render(<Info activeImage={mockActiveImage} />);
		expect(screen.getByText("John Doe")).toBeInTheDocument();
	});

	it("displays 'You' when user is the author", () => {
		render(<Info activeImage={mockActiveImage} />);
		expect(screen.getByText("You")).toBeInTheDocument();
	});

	it("displays formatted date", () => {
		render(<Info activeImage={mockActiveImage} />);
		expect(screen.getByText("20 march 2024")).toBeInTheDocument();
	});

	it("updates name when authorName changes", () => {
		const { rerender } = render(<Info activeImage={mockActiveImage} />);
		expect(screen.getByText("You")).toBeInTheDocument();

		const newImage = { ...mockActiveImage, authorName: "Jane Doe" };
		rerender(<Info activeImage={newImage} />);
		expect(screen.getByText("Jane Doe")).toBeInTheDocument();
	});

	it("updates date when dateTime changes", () => {
		const { rerender } = render(<Info activeImage={mockActiveImage} />);
		expect(screen.getByText("20 march 2024")).toBeInTheDocument();

		const newImage = { ...mockActiveImage, dateTime: "2024-03-21T10:00:00Z" };
		rerender(<Info activeImage={newImage} />);
		expect(screen.getByText("21 march 2024")).toBeInTheDocument();
	});
});

describe("Info Component Snapshot Tests", () => {
	const mockActiveImage = {
		src: "test-image.jpg",
		alt: "Test Image",
		authorName: "John Doe",
		dateTime: "2024-03-20T10:00:00Z",
		authorAvatarUrl: "avatar.jpg",
	};

	beforeEach(() => {
		mockUserStore.getState().user = {
			id: "123",
			fullName: "John Doe",
		} as IUser;
	});

	it("should match snapshot when user is the author", () => {
		const { container } = render(<Info activeImage={mockActiveImage} />);
		expect(container).toMatchSnapshot();
	});

	it("should match snapshot when user is not the author", () => {
		mockUserStore.getState().user = {
			id: "456",
			fullName: "Different User",
		} as IUser;

		const { container } = render(<Info activeImage={mockActiveImage} />);
		expect(container).toMatchSnapshot();
	});

	it("should match snapshot with different author name", () => {
		const newImage = { ...mockActiveImage, authorName: "Jane Doe" };
		const { container } = render(<Info activeImage={newImage} />);
		expect(container).toMatchSnapshot();
	});

	it("should match snapshot with different date", () => {
		const newImage = { ...mockActiveImage, dateTime: "2024-03-21T10:00:00Z" };
		const { container } = render(<Info activeImage={newImage} />);
		expect(container).toMatchSnapshot();
	});
});
