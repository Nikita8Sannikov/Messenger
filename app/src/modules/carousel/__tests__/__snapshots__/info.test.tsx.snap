// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Info Component Snapshot Tests should match snapshot when user is not the author 1`] = `
<div>
  <div
    class="info"
  >
    <span
      class="MuiBadge-root avatar-badge css-63ezxx-MuiBadge-root"
    >
      <div
        class="MuiAvatar-root MuiAvatar-circular avatar user-avatar css-fmeqll-MuiAvatar-root"
        id="user-avatar"
      >
        <img
          alt="Test Image"
          class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
          src="avatar.jpg"
        />
      </div>
      <span
        class="MuiBadge-badge MuiBadge-dot MuiBadge-anchorOriginBottomRight MuiBadge-anchorOriginBottomRightCircular MuiBadge-overlapCircular css-aa1yeu-MuiBadge-badge"
      />
    </span>
    <div
      class="info__container"
    >
      <div
        class="info__container__user-name"
      >
        John Doe
      </div>
      <div
        class="info__container__date-time"
      >
        20 march 2024
      </div>
    </div>
  </div>
</div>
`;

exports[`Info Component Snapshot Tests should match snapshot when user is the author 1`] = `
<div>
  <div
    class="info"
  >
    <span
      class="MuiBadge-root avatar-badge css-63ezxx-MuiBadge-root"
    >
      <div
        class="MuiAvatar-root MuiAvatar-circular avatar user-avatar css-fmeqll-MuiAvatar-root"
        id="user-avatar"
      >
        <img
          alt="Test Image"
          class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
          src="avatar.jpg"
        />
      </div>
      <span
        class="MuiBadge-badge MuiBadge-dot MuiBadge-anchorOriginBottomRight MuiBadge-anchorOriginBottomRightCircular MuiBadge-overlapCircular css-aa1yeu-MuiBadge-badge"
      />
    </span>
    <div
      class="info__container"
    >
      <div
        class="info__container__user-name"
      >
        You
      </div>
      <div
        class="info__container__date-time"
      >
        20 march 2024
      </div>
    </div>
  </div>
</div>
`;

exports[`Info Component Snapshot Tests should match snapshot with different author name 1`] = `
<div>
  <div
    class="info"
  >
    <span
      class="MuiBadge-root avatar-badge css-63ezxx-MuiBadge-root"
    >
      <div
        class="MuiAvatar-root MuiAvatar-circular avatar user-avatar css-fmeqll-MuiAvatar-root"
        id="user-avatar"
      >
        <img
          alt="Test Image"
          class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
          src="avatar.jpg"
        />
      </div>
      <span
        class="MuiBadge-badge MuiBadge-dot MuiBadge-anchorOriginBottomRight MuiBadge-anchorOriginBottomRightCircular MuiBadge-overlapCircular css-aa1yeu-MuiBadge-badge"
      />
    </span>
    <div
      class="info__container"
    >
      <div
        class="info__container__user-name"
      >
        Jane Doe
      </div>
      <div
        class="info__container__date-time"
      >
        20 march 2024
      </div>
    </div>
  </div>
</div>
`;

exports[`Info Component Snapshot Tests should match snapshot with different date 1`] = `
<div>
  <div
    class="info"
  >
    <span
      class="MuiBadge-root avatar-badge css-63ezxx-MuiBadge-root"
    >
      <div
        class="MuiAvatar-root MuiAvatar-circular avatar user-avatar css-fmeqll-MuiAvatar-root"
        id="user-avatar"
      >
        <img
          alt="Test Image"
          class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
          src="avatar.jpg"
        />
      </div>
      <span
        class="MuiBadge-badge MuiBadge-dot MuiBadge-anchorOriginBottomRight MuiBadge-anchorOriginBottomRightCircular MuiBadge-overlapCircular css-aa1yeu-MuiBadge-badge"
      />
    </span>
    <div
      class="info__container"
    >
      <div
        class="info__container__user-name"
      >
        You
      </div>
      <div
        class="info__container__date-time"
      >
        21 march 2024
      </div>
    </div>
  </div>
</div>
`;
